src/cluster_slot.c

```diff
@@ -55,53 +122,68 @@ static int slotStatEntryDescCmp(const void *a, const void *b) {
     return entry_b.stat - entry_a.stat;
 }
 
-static int checkSlotAssignment(client *c, unsigned char *slots, int start_slot, int end_slot) {
-    for (int slot = start_slot; slot <= end_slot; slot++) {
-        if (server.cluster->slots[slot]) {
-            if (slots[slot]++ == 1) {
-                addReplyErrorFormat(c,"Slot %d specified multiple times",(int)slot);
-                return C_ERR;
-            }
+static void sortSlotStats(sortedSlotStatEntry sorted[], int order_by, int desc) {
+    for (int slot = 0; slot < CLUSTER_SLOTS; slot++) {
+        if (doesSlotBelongToCurrentShard(slot)) {
+            sorted[slot].slot = slot;
+            sorted[slot].stat = getSingleSlotStat(slot, order_by);
+        } else {
+            /* Even if the slot does not belong to the current shard,
+             * we should fill the entry with default values so that qsort() does not segfault.
+             * These entries will be filtered and ignored upon addReplySortedSlotStats(). */
+            sorted[slot].slot = DEFAULT_SLOT;
+            sorted[slot].stat = DEFAULT_STAT;
         }
     }
-    return C_OK;
+    qsort(sorted, CLUSTER_SLOTS, sizeof(sortedSlotStatEntry), (desc) ? slotStatEntryDescCmp : slotStatEntryAscCmp);
```

Comment:
By only adding slots that are "valid" (belong to the cluster & meet the filters) we can pass the number of slots added to qsort instead, sorting on a smaller subset rather than `CLUSTER_SLOTS`. 

for example something like:

```
int res = 0;
for (int slot = 0; slot < CLUSTER_SLOTS; slot++) {
  if (doesSlotBelongToCurrentShard(slot)) {
     sorted[slot].slot = slot;
     sorted[slot].stat = getSingleSlotStat(slot, order_by);
     res++
  }
}
qsort(sorted, res, sizeof(sortedSlotStatEntry), (desc) ? slotStatEntryDescCmp : slotStatEntryAscCmp);
```